### PR TITLE
EN-2988, cookie v3 risk v2

### PR DESCRIPTION
**Pull request by nitzanpx**

### CODE REVIEW

File: perimeterx-python-wsgi.iml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (yaronpx) [inline perimeterx-python-wsgi.iml]: I don't think this file is necessary in the repository

File: perimeterx/px_cookie.py [hunk 2]
```diff
@@ -1,201 +1,178 @@
...
+    @staticmethod
+    def build_px_cookie(ctx, config):
+        config["logger"].debug("PxCookie[build_px_cookie]")
+        px_cookies = ctx['px_cookies'].keys()
+
+        # Check that its not empty
+        if not px_cookies:
+            return None
+
+        prefix = px_cookies[0]
```
Comment (yaronpx) [inline perimeterx/px_cookie.py]: Is it guaranteed that if you have both v1 and v3, the v3 will be the one you get here?

File: perimeterx/px_cookie.py [hunk 3]
```diff
@@ -1,201 +1,178 @@
...
+            parts = self.raw_cookie.split(':', 3)
+            if len(parts) != 3:
+                return False
+            salt = base64.b64decode(parts[0])
+            iterations = int(parts[1])
+            if iterations < 1 or iterations > 10000:
+                return False
+            data = base64.b64decode(parts[2])
+            dk = self.pbkdf2_hmac('sha256', self.config['cookie_key'], salt, iterations, dklen=48)
+            key = dk[:32]
```
Comment (yaronpx) [inline perimeterx/px_cookie.py]: While you're refactoring, it might be a good idea to move some of these magic numbers to the new px_constants file

**Comment by yaronpx** [comment]
What about supporting JS challenge?

File: perimeterx/px_cookie.py (hunk 2 above)
Comment (nitzanpx) [inline perimeterx/px_cookie.py]: yes, tested it

**Comment by nitzanpx** [comment]
@yaronpx 
not in this task

File: perimeterx/px_cookie.py (hunk 3 above)
Comment (nitzanpx) [inline perimeterx/px_cookie.py]: which of these magics you refer to?

**Comment by nitzanpx** [comment]
@yaronpx  please re-review

File: perimeterx/px_cookie.py (hunk 2 above)
Comment (yaronpx) [inline perimeterx/px_cookie.py]: That's not what I'm asking. The fact that it once ran and worked, does not mean it's *guaranteed* to work. 
What is the mechanism that enforces it? I didn't see any sorting of the `px_cookies` list.

File: perimeterx/px_cookie.py [hunk 4]
```diff
@@ -1,201 +1,179 @@
...
+class PxCookie:
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def build_px_cookie(ctx, config):
+        config["logger"].debug("PxCookie[build_px_cookie]")
+        px_cookies = ctx['px_cookies'].keys()
+        px_cookies.sort(reverse=True)
```
Comment (yaronpx) [inline perimeterx/px_cookie.py]: Probably better to sort *after* you check the list exists. 
Actually it's best to sort right before you store it in `px_context`

File: perimeterx/px_cookie.py (hunk 2 above)
Comment (nitzanpx) [inline perimeterx/px_cookie.py]: fixed

**Pull request merged**